lib/Parse/ParseDecl.cpp

```diff
@@ -5944,14 +5952,21 @@ void Parser::ParseParameterDeclarationClause(
       // added to the current scope.
       ParmVarDecl *Param = Actions.ActOnParamDeclarator(getCurScope(), ParmDeclarator);
 
-      // Parse the bounds expression, if any.
+      // Parse and store the tokens for the bounds expression, if there is one.
       if (getLangOpts().CheckedC && Tok.is(tok::colon)) {
         ConsumeToken();
-        ExprResult Bounds = ParseBoundsExpression();
-        if (Bounds.isInvalid()) {
+        // Consume and store tokens until a bounds-like expression has been
+        // read or a parsing error has happened.  Store the tokens even if a
+        // parsing error occurs so that ParseBoundsExpression can generate the
+        // error message.  This way the error messages from parsing of bounds
+        // expressions will be the same or very similar regardless of whether
+        // parsing is deferred or not.
+        CachedTokens *BoundsExprTokens = new CachedTokens;
```

Comment:
clang and LLVM use an explicit error code/value style, instead of C++ exceptions.    I couldn't find documentation on which smart pointer class they want to use.  I'll add a FIXME comment like there is in the similar code below.